corehq/ex-submodules/couchforms/models.py

```diff
@@ -346,27 +346,54 @@ def _meta_to_json(meta):
     def xml_md5(self):
         return hashlib.md5(self.get_xml().encode('utf-8')).hexdigest()
 
-    def archive(self, user_id=None):
-        if self.is_archived:
-            return
-        self.doc_type = "XFormArchived"
-        self.history.append(XFormOperation(
-            user=user_id,
-            operation='archive',
-        ))
-        self.save()
-        xform_archived.send(sender="couchforms", xform=self)
+    def archive(self, user_id=None, retry_archive=False):
```

Comment:
I think you can get rid of the `retry_archive` kwarg and related branching

Reply:
I think it should stay for the case where the doctype successfully updated but nothing else went through. If that's the case, then on retry (for an archive), it'll see that `self.is_archived` is True and it'll skip the rest of the logic.

Reply:
I think we're missing each other a bit here. If we're updating the stub after the doc_type and history are saved then we know that they have been updated:
```
with stub:
    form.doc_type = 'X'
    form.save()
    stub.archive_history_updated()
    signal.send(...)
```
I noticed below that you were calling `stub.archive_history_updated()` before the form was saved.